hivemq-edge/src/main/java/com/hivemq/combining/CombiningManager.java

```diff
@@ -0,0 +1,150 @@
+package com.hivemq.combining;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.base.Preconditions;
+import com.hivemq.adapter.sdk.api.events.EventService;
+import com.hivemq.adapter.sdk.api.events.model.Event;
+import com.hivemq.persistence.generic.AddResult;
+import com.hivemq.protocols.ConfigPersistence;
+import com.hivemq.protocols.ProtocolAdapterConfig;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+
+import static com.hivemq.persistence.util.FutureUtils.syncFuture;
+
+@Singleton
+public class CombiningManager {
+
+    private static final Logger log = LoggerFactory.getLogger(CombiningManager.class);
+
+    private final @NotNull ConfigPersistence configPersistence;
+    private final @NotNull EventService eventService;
+    private final @NotNull MetricRegistry metricRegistry;
+    private final Map<String, DataCombiner> idToDataCombiner = new ConcurrentHashMap<>();
+
+    @Inject
+    public CombiningManager(
+            final @NotNull ConfigPersistence configPersistence,
+            final @NotNull EventService eventService,
+            final @NotNull MetricRegistry metricRegistry) {
+        this.configPersistence = configPersistence;
+        this.eventService = eventService;
+        this.metricRegistry = metricRegistry;
+        // TODO move to HiveMQMetrics
+        metricRegistry.registerGauge("com.hivemq.edge.data-combining.data-combiners.current",
+                () -> idToDataCombiner.values().size());
+    }
+
+
+    public void start() {
+        log.debug("Starting data combiners");
+        for (final ProtocolAdapterConfig adapterConfig : configPersistence.allAdapters()) {
+
+            // TODO start the actual data combining
+        }
+    }
+
+    public @NotNull CompletableFuture<AddResult> startCombiner(final @NotNull DataCombiner dataCombiner) {
+        log.debug("Starting data combiners with id '{}'", dataCombiner.id());
+        // TODO
+        return CompletableFuture.completedFuture(AddResult.success());
+    }
+
+    public @NotNull CompletableFuture<Void> stop(final @NotNull String dataCombinerId) {
+        // TODO
+        return CompletableFuture.completedFuture(null);
+    }
+
+    public @NotNull CompletableFuture<Void> stopAll() {
+        // TODO
+        return CompletableFuture.completedFuture(null);
+    }
+
+    public @NotNull CompletableFuture<AddResult> addDataCombiner(final @NotNull DataCombiner dataCombiner) {
+        if (getCombinerById(dataCombiner.id()).isPresent()) {
```

Comment:
A saver way is to use putIfAbsent. It returns null if the key wasn't defined before and you also stay within the guarantees for atomicity of a concurrenthashmap.